### PR TITLE
fix(sdk-middleware-http): allow to unset content-type

### PR DESCRIPTION
**Pull request by MicheleRezk**
#### Summary
Don't set the content-type by default to the requests with Null content type, and let the client to identify the correct content-type based on the request body

#### Description
Adjust the http middleware to check for requests with Null Content-Type, if this passed in the request headers then we don't need to set the content-type for the request to be 'application/json' and let the client identify the correct content-type based on the request body.

This is useful when the request is of type 'multipart/form-data' and it's better to let the client Identify the correct content-type & boundary and also calculate the Content-Length. 

so while calling file-upload service we will pass this Content-Type as Null in the request headers without identifying the content-type and the sdk client will add the correct content-type while calling the service.

- Tests
  - [ ] Unit

### CODE REVIEW

**Comment by tdeekens** [comment]
Thanks for the changes @MicheleRezk! I can follow along the concepts however it feels diffs contain a lot of changes due to intention. Can you maybe perform a self review and highlight and explain important changes? That would ease and speed up the review. Thanks.

**Comment by MicheleRezk** [comment]
sorry @tdeekens, I adjusted the file format and spaces, it should be easier now for review, please let me know if you have any other comment, thanks a lot

**Comment by emmenko** [comment]
Quick feedback: I wouldn't use a custom header to disable another header.

I would implement it so that you pass `null` as the `Content-Type` header and the middleware then unsets the header when `null` is passed.

File: packages/sdk-middleware-http/src/http.js [hunk 1]
```diff
@@ -100,17 +100,25 @@ export default function createHttpMiddleware({
...
     let abortController: any
     const url = host.replace(/\/$/, '') + request.uri
+
+    // if it's true, the client will detect the correct content-type, will be used in sending files (ex: multipart/form-data)
+    const hasNullContentType = request.headers["Content-Type"] === null
     const body =
-      typeof request.body === 'string' || Buffer.isBuffer(request.body)
+      typeof request.body === 'string' ||
+        Buffer.isBuffer(request.body) ||
+        hasNullContentType
```
Comment (emmenko) [inline packages/sdk-middleware-http/src/http.js:109]:
```suggestion
        request.body instanceof FormData
```

**Review by tdeekens — APPROVED** [review]
From what I can tell and knowledge of the goal and topic this implementation should do the job.

File: packages/sdk-middleware-http/src/http.js [hunk 2]
```diff
@@ -100,17 +100,25 @@ export default function createHttpMiddleware({
...
+        JSON.stringify(request.body || undefined)
 
     const requestHeader: HttpHeaders = { ...request.headers }
     if (!Object.prototype.hasOwnProperty.call(requestHeader, 'Content-Type')) {
       requestHeader['Content-Type'] = 'application/json'
     }
-    if (body) {
+    if(hasNullContentType) {
+      delete requestHeader['Content-Type']
+    }
```
Comment (emmenko) [inline packages/sdk-middleware-http/src/http.js:120]:
```suggestion
    if (requestHeader['Content-Type'] === null) {
      delete requestHeader['Content-Type']
    }
```

File: packages/sdk-middleware-http/src/http.js [hunk 3]
```diff
@@ -100,17 +100,25 @@ export default function createHttpMiddleware({
...
+    else if (body) {
```
Comment (emmenko) [inline packages/sdk-middleware-http/src/http.js:121]:
```suggestion
    if (body && !(body instanceof FormData)) {
```

File: packages/sdk-middleware-http/src/http.js [hunk 4]
```diff
@@ -100,17 +100,25 @@ export default function createHttpMiddleware({
   ) => {
     let abortController: any
     const url = host.replace(/\/$/, '') + request.uri
+
+    // if it's true, the client will detect the correct content-type, will be used in sending files (ex: multipart/form-data)
+    const hasNullContentType = request.headers["Content-Type"] === null
```
Comment (emmenko) [inline packages/sdk-middleware-http/src/http.js:105]:
```suggestion
```

**Review by emmenko — CHANGES_REQUESTED** [review]
The `hasNullContentType` logic was confusing me a bit. I tried to make it a bit more explicit (suggestions).

File: packages/sdk-middleware-http/src/http.js (hunk 1 above)
Comment (MicheleRezk) [reply packages/sdk-middleware-http/src/http.js:109]: Thanks @emmenko for feedback, I didn't want to restrict it to FormData as we may need to refactor this again if we introduced a new service with application/x-www-form-urlencoded content type

File: packages/sdk-middleware-http/src/http.js [hunk 5]
```diff
@@ -101,16 +101,21 @@ export default function createHttpMiddleware({
     let abortController: any
     const url = host.replace(/\/$/, '') + request.uri
     const body =
-      typeof request.body === 'string' || Buffer.isBuffer(request.body)
+      typeof request.body === 'string' ||
+        Buffer.isBuffer(request.body) ||
+        (request.body instanceof FormData)
```
[the remark this answers is not in the document]
Comment (emmenko) [reply packages/sdk-middleware-http/src/http.js:106]: Hmm yeah that sucks...maybe we can refactor this a bit and use the `content-type` info to run the check.

For instance, only if `application/json` we need to `JSON.stringify`.

[the remark this answers is not in the document]
Comment (emmenko) [reply packages/sdk-middleware-http/src/http.js:106]:
```js
if (!Object.prototype.hasOwnProperty.call(requestHeader, 'Content-Type')) {
  requestHeader['Content-Type'] = 'application/json'
}
if(request.headers["Content-Type"] === null) {
  delete requestHeader['Content-Type']
}

const body = requestHeader['Content-Type'] === 'application/json'
  // NOTE: `stringify` of `null` gives the String('null')
  ? JSON.stringify(request.body || undefined)
  : request.body

if (typeof body === 'string' || Buffer.isBuffer(request.body)) {
  requestHeader['Content-Length'] = Buffer.byteLength(body).toString()
}
```

[the remark this answers is not in the document]
Comment (MicheleRezk) [reply packages/sdk-middleware-http/src/http.js:106]: yes I think this is much better, thanks for the feedback :+1:

[the remark this answers is not in the document]
Comment (tdeekens) [reply packages/sdk-middleware-http/src/http.js:106]: Yes, I am not 100% into this topic but wouldn't using the `content-type` header and `multipart/form-data` work?

[the remark this answers is not in the document]
Comment (MicheleRezk) [reply packages/sdk-middleware-http/src/http.js:106]: @tdeekens if we passed the content-type explicitly as 'multipart/form-data' in the request header it wouldn't work as there is other things need to be calculated like boundary, so to bypass this we can pass content-type as null and let the client Identify the correct content-type and the boundary to make it work

**Review by emmenko — APPROVED** [review]
Thanks!

**Comment by emmenko** [comment]
@MicheleRezk btw, you need to write a changeset so that the changes here can be released.

**Review by ajimae — DISMISSED** [review]
LGTM!

File: .changeset/orange-years-rush.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+"@commercetools/sdk-middleware-http": patch
+---
+
+Allow requests with empty content-type
```
Comment (emmenko) [inline .changeset/orange-years-rush.md:5]: Please specify that you need to explicitly pass `null` to "unset" the header.

Comment (ajimae) [reply .changeset/orange-years-rush.md:5]: Hi @emmenko 
This is how I worded it: To unset a request header's content-type, explicitly pass `null` as the value. 

Hope it's fine?

**Review by tdeekens — DISMISSED** [review]

File: .changeset/plenty-wasps-refuse.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+---
+"@commercetools/sdk-middleware-http": patch
+---
+
+- Allow requests with empty content-type
```
Comment (emmenko) [inline .changeset/plenty-wasps-refuse.md:5]: Remove this line. Also the other changeset file

File: integration-tests/cli/discount-codes.it.js [hunk 8]
```diff
@@ -51,6 +51,8 @@ describe('DiscountCode tests', () => {
     ])
 
     cartDiscount = data[0].body
+
+    console.log("""""""""""""""""""""""""""""""""""" + cartDiscount + """"""""""""""""""""""""""""""""""")
```
Comment (emmenko) [inline integration-tests/cli/discount-codes.it.js:55]: Remove

File: packages/sdk-middleware-http/src/http.js [hunk 9]
```diff
@@ -100,17 +100,19 @@ export default function createHttpMiddleware({
...
       requestHeader['Content-Type'] = 'application/json'
     }
-    if (body) {
+    if(request.headers["Content-Type"] === null) {
+      delete requestHeader['Content-Type']
+    }
+    const body = requestHeader['Content-Type'] === 'application/json'
+        // NOTE: `stringify` of `null` gives the String('null')
+        ? JSON.stringify(request.body || undefined)
+        : request.body
```
Comment (emmenko) [inline packages/sdk-middleware-http/src/http.js:113]: The failing integration tests might be related to this change. If you want to debug the error, print out the `console.log(body)` to see what value it is.

**Review by emmenko — CHANGES_REQUESTED** [review]

**Comment by emmenko** [comment]
I think the changes here are good now. The failing integration tests are the same as in `master`, so unrelated to these changes.

**Review by emmenko — APPROVED** [review]

**Review by ajimae — APPROVED** [review]

**Comment by emmenko** [comment]
@tdeekens anything else from your side?

**Pull request merged**